src/agent/windows/WMICommunication.cpp

```diff
@@ -0,0 +1,215 @@
+            NULL,
+            &m_pEnumerator);
+
+        if (FAILED(hres))
+        {
+            throw std::exception("Query for operating system name failed. Error code = 0x");
+        }
+
+        IWbemClassObject* pclsObj = NULL;
+        ULONG uReturn = 0;
+
+        while (m_pEnumerator)
+        {
+            HRESULT hr = m_pEnumerator->Next(WBEM_INFINITE, 1,
+                &pclsObj, &uReturn);
+
+            if (0 == uReturn)
+            {
+                break;
+            }
+
+            VARIANT vtProp;
+            hr = pclsObj->Get(L"VendorSpecific", 0, &vtProp, 0, 0);
+
+            if (V_ISARRAY(&vtProp))
+            {
+                LPSAFEARRAY pSafeArray = V_ARRAY(&vtProp);
+
+                VARTYPE itemType;
+                if (SUCCEEDED(SafeArrayGetVartype(pSafeArray, &itemType)))
+                {
+                    if (itemType == VT_UI1)
+                    {
+                        if (SafeArrayGetDim(pSafeArray) == 1)
+                        {
+                            LONG lBound;
+                            LONG uBound;
+
+                            if (SUCCEEDED(SafeArrayGetLBound(pSafeArray, 1, &lBound)) && SUCCEEDED(SafeArrayGetUBound(pSafeArray, 1, &uBound)))
+                            {
+                                const LONG itemCount = uBound - lBound + 1;
+
+                                std::vector<BYTE> dataFromArray(itemCount);
+
+                                BYTE* safearrayData;
+                                hr = SafeArrayAccessData(pSafeArray, reinterpret_cast<LPVOID*>(&safearrayData));
+                                if (FAILED(hr))
+                                {
+                                    dataFromArray.clear();
+                                }
+
+                                memcpy(dataFromArray.data(), safearrayData, itemCount);
+
+                                hr = SafeArrayUnaccessData(pSafeArray);
+                                if (FAILED(hr))
+                                {
+                                    dataFromArray.clear();
+                                }
+
+                                if (dataFromArray.empty() != true)
+                                {
+                                    FeedSmartDataStructure(dataFromArray, itemCount);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            VariantClear(&vtProp);
+
+            pclsObj->Release();
+        }
+
+        return true;
+    }
+    catch (...)
+    {
+        if (m_services != NULL)
+        {
+            m_services->Release();
+        }
+
+        if (m_initialLocatorToWMI != NULL)
+        {
+            m_initialLocatorToWMI->Release();
+        }
+
+        CoUninitialize();
+
+        return false;
+    }
+}
+
+void WMICommunication::FeedSmartDataStructure(std::vector<BYTE>& data, const LONG& dataSize)
```

Comment:
isn't data already a datavector?

Reply:
This is a method which will feed data structure of common smart data. here I feed vector but in near future here will be something else. So I will left this as is for now.